optaplanner-core/src/main/java/org/optaplanner/core/api/score/AbstractScore.java

```diff
@@ -27,16 +27,16 @@
  * <p>
  * Subclasses must be immutable.
  *
- * @param <S> the actual score type
+ * @param <Score_> the actual score type
  * @see Score
  * @see HardSoftScore
  */
-public abstract class AbstractScore<S extends Score> implements Score<S>,
+public abstract class AbstractScore<Score_ extends AbstractScore<Score_>> implements Score<Score_>,
         Serializable {
 
     protected static final String INIT_LABEL = "init";
 
-    protected static String[] parseScoreTokens(Class<? extends Score> scoreClass,
+    protected static String[] parseScoreTokens(Class<? extends Score<?>> scoreClass,
```

Comment:
`Class<? extends AbstractScore<?>>`?
I'm thinking if we should just make an instance method, and use `Score_` instead.
(Likewise for other similar methods on this class.)